### PR TITLE
[SPARK-46380][SQL]Replace current time/date prior to evaluating inline table expressions.

### PR DESCRIPTION
**Pull request by dbatomic**
### What changes were proposed in this pull request?

With this PR proposal is to do inline table resolution in two phases:
1) If there are no expressions that depend on current context (e.g. expressions that depend on CURRENT_DATABASE, CURRENT_USER, CURRENT_TIME etc.) they will be evaluated as part of ResolveInlineTable rule.
2) Expressions that do depend on CURRENT_* evaluation will be kept as expressions and they evaluation will be delayed to post analysis phase.

### Why are the changes needed?

This PR aims to solve two problems with inline tables.

Example1:
```sql
SELECT COUNT(DISTINCT ct) FROM VALUES
(CURRENT_TIMESTAMP()),
(CURRENT_TIMESTAMP()),
(CURRENT_TIMESTAMP()) as data(ct)
```
Prior to this change this example would return 3 (i.e. all CURRENT_TIMESTAMP expressions would return different value since they would be evaluated individually as part of inline table evaluation). After this change result is 1.

Example 2:
```sql
CREATE VIEW V as (SELECT * FROM VALUES(CURRENT_TIMESTAMP())
```
In this example VIEW would be saved with literal evaluated during VIEW creation. After this change CURRENT_TIMESTAMP() will eval during VIEW execution.

### Does this PR introduce _any_ user-facing change?

See section above.

### How was this patch tested?

New test that validates this behaviour is introduced.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala [hunk 1]
```diff
@@ -33,12 +34,14 @@ import org.apache.spark.sql.types.{StructField, StructType}
...
 object ResolveInlineTables extends Rule[LogicalPlan]
   with CastSupport with AliasHelper with EvalHelper {
-  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
-    AlwaysProcess.fn, ruleId) {
-    case table: UnresolvedInlineTable if table.expressionsResolved =>
-      validateInputDimension(table)
-      validateInputEvaluable(table)
-      convert(table)
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.resolveOperatorsDown(p => ComputeCurrentTime(p)).resolveOperatorsWithPruning(
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: Just wonder, can't we just move `ComputeCurrentTime` earlier somewhere and skip resolution here if there is the pattern `CURRENT_LIKE`?

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: I thought about the same thing but there is this comment:

    // Technically some of the rules in Finish Analysis are not optimizer rules and belong more
    // in the analyzer, because they are needed for correctness (e.g. ComputeCurrentTime).
    // However, because we also use the analyzer to canonicalized queries (for view definition),
    // we do not eliminate subqueries or compute current time in the analyzer.
    
I also wandered why we don't do ComputeCurrentTime in analysis phase, but this comment does make sense.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: `ComputeCurrentTime` transforms the entire plan tree, so the code should be `ComputeCurrentTime(plan).resolveOperatorsWithPruning ...`

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: I did try that originally but I got:
`      throw QueryExecutionErrors.methodCalledInAnalyzerNotAllowedError()` 

I ended up with resolveOperatorsDown due to this comment:
```scala
  /**
   * In analyzer, use [[resolveOperatorsDown()]] instead. If this is used in the analyzer,
   * an exception will be thrown in test mode. It is however OK to call this function within
   * the scope of a [[resolveOperatorsDown()]] call.
   * @see [[org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning()]].
   */
  override def transformDownWithPruning(cond: TreePatternBits => Boolean,

```

**Review by beliefer — COMMENTED** [review]
I think we need a discussion.
```
SELECT COUNT(DISTINCT ct) FROM VALUES
(CURRENT_TIMESTAMP()),
(CURRENT_TIMESTAMP()),
(CURRENT_TIMESTAMP()) as data(ct)
```
The three call of CURRENT_TIMESTAMP() should have the same value?

**Comment by dbatomic** [comment]
> I think we need a discussion.
> 
> ```
> SELECT COUNT(DISTINCT ct) FROM VALUES
> (CURRENT_TIMESTAMP()),
> (CURRENT_TIMESTAMP()),
> (CURRENT_TIMESTAMP()) as data(ct)
> ```
> 
> The three call of CURRENT_TIMESTAMP() should have the same value?

That's right. All the invocations of CURRENT_TIMESTAMP/CURRENT_DATE/NOW() should be replaced with single value which represents time of query arrival. We already do this for majority of scenarios (e.g. if time function is pretty much anywhere else in the query). The bug this PR tries to solve is that we don't do this for inline tables.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: We can wrap the code with `AnalysisHelper.allowInvokingTransformsInAnalyzer` to bypass this limitation.

**Comment by beliefer** [comment]
I checked in Postgres.
`select CURRENT_TIMESTAMP, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP;`
The output:
```
current_timestamp            |current_timestamp            |current_timestamp            |
-----------------------------+-----------------------------+-----------------------------+
2023-12-14 16:27:19.663 +0800|2023-12-14 16:27:19.663 +0800|2023-12-14 16:27:19.663 +0800|
```

**Comment by dbatomic** [comment]
> I checked in Postgres. `select CURRENT_TIMESTAMP, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP;` The output:
> 
> ```
> current_timestamp            |current_timestamp            |current_timestamp            |
> -----------------------------+-----------------------------+-----------------------------+
> 2023-12-14 16:27:19.663 +0800|2023-12-14 16:27:19.663 +0800|2023-12-14 16:27:19.663 +0800|
> ```

Yep, just to illustrate that. the same works for inline tables:

```sql
postgres=# SELECT * FROM (VALUES(EXTRACT(epoch FROM current_timestamp),EXTRACT(epoch FROM current_timestamp),EXTRACT(epoch FROM current_timestamp)));
```

      column1      |      column2      |      column3
-------------------+-------------------+-------------------
 1702546463.398428 | 1702546463.398428 | 1702546463.398428

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala (hunk 1 above)
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: I think that my current approach is not valid. i.e. if we do ComputeCurrentTime here we will break following assumption:

// Technically some of the rules in Finish Analysis are not optimizer rules and belong more
// in the analyzer, because they are needed for correctness (e.g. ComputeCurrentTime).
// However, because we also use the analyzer to canonicalized queries (for view definition),
// we do not eliminate subqueries or compute current time in the analyzer.

So, if we had inline table in view definition this wouldn't work (although this is currently incorrect as well). 

On the other hand, if we were to replace current time only for inline table values then we might end up with different results for CURRENT_DATE/TIMESTAMP in other parts of the plan so that doesn't work either.

I will see if I can delay inline table evaluation, or at least delay it up for expressions for which we can guarantee that they will be replaced by a literal. This PR is on hold until then.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:38]: Oh good point. We should have a logical plan for inline tables to keep the expressions, and then replace it with `LocalRelation` after we run `ComputeCurrentTime`

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala]: I played a bit more and came up with solution that is in current PR (which I don't like that much to be honest, since it adds another rule in post analyzer phase).

Here is the rough logic:
1) In `ResolveInlineTable` we make sure that shape is correct and that there is proper casting for all the objects. Here we also decide whether all expressions can be reduced to `LocalRelation` or we need to postpone that step. Option here is to always postpone - I originally went with that approach but that changed bunch of GOLDEN files since we now have TempResolvedInlineTable after every analysis run. But still, it is cleaner to always postpone - looking for feedback here.
2) In EvalInlineTable, after ReplaceCurrentTime/Like we do final evaluation if needed.

By doing this we can solve a couple of additional "bugs" - e.g. when creating view as CREATE VIEW V AS (SELECT * FROM VALUES (NOW())) previous behaviour was to embed time of view creation in query, while proper thing should be to calculate time on every invocation of the view (I verified that PG does this).

An option is to also do everything in `EvalInlineTable` rule. The problem here is that I think we will want at least to check whether query is correct in analysis phase (i.e. check casting and shape) and then do final eval() in finishAnalysis phase, which is part of optimizer.

In short, we should have short discussion about this. @cloud-fan , @MaxGekk, @beliefer, @dtenedor and @srielau - please provide feedback.

**Comment by srielau** [comment]
I agree with the semantic changes. General comment, though:
VALUES in its current implementation is much too restrictive.
Will this PR move us closer to allowing arbitrary expressions (including non determinism and correlation)?
E.g. the following cannot be done in VALUES today:
scala> spark.sql("VALUES (rand())").show()
org.apache.spark.sql.AnalysisException: [INVALID_INLINE_TABLE.CANNOT_EVALUATE_EXPRESSION_IN_INLINE_TABLE] Invalid inline table. Cannot evaluate the expression "rand()" in inline table definition. SQLSTATE: 42000; line 1 pos 8

or:
SELECT pk, c FROM T, LATERAL(VALUES(T.c1), (T.c2)) AS unpivot(c)

**Comment by dbatomic** [comment]
> I agree with the semantic changes. General comment, though: VALUES in its current implementation is much too restrictive. Will this PR move us closer to allowing arbitrary expressions (including non determinism and correlation)? E.g. the following cannot be done in VALUES today: scala> spark.sql("VALUES (rand())").show() org.apache.spark.sql.AnalysisException: [INVALID_INLINE_TABLE.CANNOT_EVALUATE_EXPRESSION_IN_INLINE_TABLE] Invalid inline table. Cannot evaluate the expression "rand()" in inline table definition. SQLSTATE: 42000; line 1 pos 8
> 
> or: SELECT pk, c FROM T, LATERAL(VALUES(T.c1), (T.c2)) AS unpivot(c)

Yeah, I maybe can use this PR to deal with rand() problem as well (i.e. allow non-deterministic expressions in VALUES), because it is rather similar to CURRENT_* issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 2]
```diff
@@ -128,6 +128,19 @@ case class UnresolvedInlineTable(
...
+ * An temp resolved inline table that holds all the expressions that were checked for
+ * the right shape and common data types.
+ * This is a preparation step for [[org.apache.spark.sql.catalyst.optimizer.EvalInlineTables]] which
+ * will produce a [[org.apache.spark.sql.catalyst.plans.logical.LocalRelation]]
+ * for this inline table.
+ *
+ * @param output list of column attributes
+ * @param rows expressions for the data rows
+ */
+case class TempResolvedInlineTable(rows: Seq[Seq[Expression]], output: Seq[Attribute])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:141]: it's fully resolved, just not evaluated. Maybe remove tmp from the name?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:141]: To avoid any perf regression, we can put `rows` in a wrapper class so that catalyst rules won't see them by default. These expressions do not need any more extra resolution.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:141]: maybe this is over-conservative. Many rules will be skipped if the expression is already resolved. We can add the wrapper class when needed in the future.

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Yep, I also don't think it is a good idea to be that conservative here. Especially if we want to move in direction of making inline tables more flexible, as Serge pointed out.

Please take a look at the latest iteration.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala [hunk 3]
```diff
@@ -105,26 +107,40 @@ object ResolveInlineTables extends Rule[LogicalPlan]
...
-        }
-      })
+          castedExpr
+      }
     }
 
-    LocalRelation(attributes, newRows)
+    val noCurrentLike: Boolean = castedRows.flatten.forall(!_.containsPattern(CURRENT_LIKE))
+    if (noCurrentLike) {
+      val evalRows: Seq[InternalRow] =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:125]: to avoid duplicated code, shall we always create `ResolvedInlineTable` and then invoke `EvalInlineTables.apply` when there is no current_like?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 4]
```diff
@@ -128,6 +128,19 @@ case class UnresolvedInlineTable(
...
+ * the right shape and common data types.
+ * This is a preparation step for [[org.apache.spark.sql.catalyst.optimizer.EvalInlineTables]] which
+ * will produce a [[org.apache.spark.sql.catalyst.plans.logical.LocalRelation]]
+ * for this inline table.
+ *
+ * @param output list of column attributes
+ * @param rows expressions for the data rows
+ */
+case class ResolvedInlineTable(rows: Seq[Seq[Expression]], output: Seq[Attribute])
+  extends LeafNode { }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:142]:
```suggestion
  extends LeafNode
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 5]
```diff
@@ -70,6 +74,33 @@ object RewriteNonCorrelatedExists extends Rule[LogicalPlan] {
...
+ * of the analysis phase, given that all the expressions need to be fully resolved/replaced
+ * at this point.
+ */
+object EvalInlineTables extends Rule[LogicalPlan] with CastSupport {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformDownWithSubqueriesAndPruning(
+    AlwaysProcess.fn, ruleId) {
+    case table: ResolvedInlineTable =>
+      val newRows: Seq[InternalRow] =
+        table.rows.map { row => InternalRow.fromSeq(
+            row.map(e =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:88]: nit: for multi-line lambda we should wrap with `row.map { e => ...`

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 6]
```diff
@@ -78,6 +78,11 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
 
     // Execution in one query should return the same value
     checkAnswer(sql("""SELECT CURRENT_TIMESTAMP() = CURRENT_TIMESTAMP()"""), Row(true))
+    checkAnswer(sql(
+      """SELECT COUNT(DISTINCT ct) FROM VALUES
+        | CURRENT_TIMESTAMP(),
+        | CURRENT_TIMESTAMP(),
+        | CURRENT_TIMESTAMP() as data(ct)""".stripMargin), Row(1))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala:85]: This seems not needed as we have the same test in the golden file

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala (hunk 3 above)
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala]: Yep, good idea. I also organized code a bit better and made it more testable.

File: sql/core/src/test/resources/sql-tests/results/inline-table.sql.out [hunk 7]
```diff
@@ -271,3 +271,11 @@ select * from values (10 + try_divide(5, 0))
 struct<col1:double>
 -- !query output
 NULL
+
+
+-- !query
+select count(distinct ct) from values now(), now(), now() as data(ct)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/inline-table.sql.out:277]: We can add a test for `CURRENT_TIMESTAMP` and remove https://github.com/apache/spark/pull/44316/files#r1430980065

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala [hunk 8]
```diff
@@ -105,26 +107,31 @@ object ResolveInlineTables extends Rule[LogicalPlan]
...
+  /**
+   * This function attempts to early evaluate rows in inline table.
+   * If evaluation doesn't rely on non-deterministic expressions (e.g. current_like)
+   * expressions will be evaluated and inlined as [[LocalRelation]]
+   * This is package visible for unit testing.
+   */
+  private[analysis] def earlyEvalIfPossible(table: ResolvedInlineTable): LogicalPlan = {
+      def earlyEvalPossible =
+        table.rows.flatten.forall(!_.containsPattern(CURRENT_LIKE))
+      if (earlyEvalPossible) EvalInlineTables(table) else table
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:135]:
```suggestion
    val earlyEvalPossible = table.rows.flatten.forall(!_.containsPattern(CURRENT_LIKE))
    if (earlyEvalPossible) EvalInlineTables(table) else table
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 9]
```diff
@@ -70,6 +75,33 @@ object RewriteNonCorrelatedExists extends Rule[LogicalPlan] {
...
+ * of the analysis phase, given that all the expressions need to be fully resolved/replaced
+ * at this point.
+ */
+object EvalInlineTables extends Rule[LogicalPlan] with CastSupport {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformDownWithSubqueriesAndPruning(
+    AlwaysProcess.fn, ruleId) {
+    case table: ResolvedInlineTable =>
+      val newRows: Seq[InternalRow] =
+        table.rows.map { row => InternalRow.fromSeq(
+            row.map { e =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:89]: nit: 2 spaces indentation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 10]
```diff
@@ -70,6 +75,33 @@ object RewriteNonCorrelatedExists extends Rule[LogicalPlan] {
...
 }
 
+/**
+ * Computes expressions in inline tables. This rule is supposed to be called at the very end
+ * of the analysis phase, given that all the expressions need to be fully resolved/replaced
+ * at this point.
+ */
+object EvalInlineTables extends Rule[LogicalPlan] with CastSupport {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformDownWithSubqueriesAndPruning(
+    AlwaysProcess.fn, ruleId) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:85]: let's add pruning for `ResolvedInlineTable`

File: sql/core/src/test/resources/sql-tests/inputs/inline-table.sql [hunk 11]
```diff
@@ -60,3 +60,9 @@ select * from values (timestamp('1991-12-06 00:00:00.0'), array(timestamp('1991-
 select * from values (try_add(5, 0));
 select * from values (try_divide(5, 0));
 select * from values (10 + try_divide(5, 0));
+
+-- now() should be kept as tempResolved inline expression.
+select count(distinct ct) from values now(), now(), now() as data(ct);
+
+-- current_timestamp() should be kept as tempResolved inline expression.
+select count(distinct ct) from values current_timestamp(), current_timestamp() as data(ct);
```
Comment (beliefer) [inline sql/core/src/test/resources/sql-tests/inputs/inline-table.sql:68]: Shall we add tests mixed `current_timestamp` and other deterministic function?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 12]
```diff
@@ -70,6 +75,34 @@ object RewriteNonCorrelatedExists extends Rule[LogicalPlan] {
...
+ * of the analysis phase, given that all the expressions need to be fully resolved/replaced
+ * at this point.
+ */
+object EvalInlineTables extends Rule[LogicalPlan] with CastSupport {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.transformDownWithSubqueriesAndPruning(_.containsPattern(INLINE_TABLE_EVAL)) {
+      case table: ResolvedInlineTable =>
+        val newRows: Seq[InternalRow] =
+          table.rows.map { row => InternalRow.fromSeq(
+            row.map { e =>
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:89]:
```suggestion
          table.rows.map { row => InternalRow.fromSeq(row.map { e =>
```

**Review by beliefer — APPROVED** [review]
LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala [hunk 13]
```diff
@@ -105,26 +107,30 @@ object ResolveInlineTables extends Rule[LogicalPlan]
     val attributes = DataTypeUtils.toAttributes(StructType(fields))
     assert(fields.size == table.names.size)
 
-    val newRows: Seq[InternalRow] = table.rows.map { row =>
-      InternalRow.fromSeq(row.zipWithIndex.map { case (e, ci) =>
-        val targetType = fields(ci).dataType
-        try {
+    val castedRows: Seq[Seq[Expression]] = table.rows.map { row =>
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:110]: It seems we only need the `Seq[Expression]` here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 14]
```diff
@@ -131,6 +131,21 @@ case class UnresolvedInlineTable(
...
+ * An resolved inline table that holds all the expressions that were checked for
+ * the right shape and common data types.
+ * This is a preparation step for [[org.apache.spark.sql.catalyst.optimizer.EvalInlineTables]] which
+ * will produce a [[org.apache.spark.sql.catalyst.plans.logical.LocalRelation]]
+ * for this inline table.
+ *
+ * @param output list of column attributes
+ * @param rows expressions for the data rows
+ */
+case class ResolvedInlineTable(rows: Seq[Seq[Expression]], output: Seq[Attribute])
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:144]: Shall we simplify `rows: Seq[Seq[Expression]]` to `exprs: Seq[Expression]`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala (hunk 13 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:110]: it's a table (rows X columns)

File: sql/core/src/test/resources/sql-tests/inputs/inline-table.sql (hunk 11 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/inline-table.sql:68]: it's testing the correct value using count distinct.

**Comment by cloud-fan** [comment]
The failed test is unrelated, thanks, merging to master/3.5!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala (hunk 13 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:110]: I got it now. Thank you!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala (hunk 14 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:144]: @dbatomic After review this PR again. I'm sorry for the above comment.
